# Structured Programming

## Lab 8. Subprograms and Recursion

### Learning Objectives

By the end of this worksheet, you will be able to:

* Differentiate between functions and procedures.
* Identify the basic structure of subprograms.
* Create functions and procedures in Python.
* Understand the concept of recursion.
* Implement simple recursive functions.



# Part 1. Reviewing Functions

Observe the following code and execute it.



In [ ]:
def calculate_final_grade(exam1, exam2, project):
  return (exam1 + exam2 + project) / 3

final_grade = calculate_final_grade(85, 92, 88)
print(final_grade)

88.33333333333333




### Reflection Questions

1. What value does the function return?

# The function returns the calculated average of the three inputs: 88.333
---

2. Where is the result stored?


# The result is stored in the variable named final_grade

---

3. What would happen if the `return` statement were removed?

# The function would still execute the calculation, but it would not send any value back to the main program. As a result, the variable final_grade would be assigned the value None, and printing it would display None.

---



# Part 2. What Is a Procedure?

In structured programming, there are two main types of subprograms:

## Function

A function performs a task and returns a value.

Example:



In [ ]:
def calculate_area(base, height):
    return base * height

area = calculate_area(5, 4)

print(area)

20




## Procedure

A procedure performs an action but does not return a value.

Example:



In [ ]:
def display_menu():
    print("1. Add")
    print("2. Subtract")
    print("3. Exit")

display_menu()

1. Add
2. Subtract
3. Exit




### Activity 1

Execute the following code.



In [ ]:
def greet(name):
    print("Hello", name)

result = greet("Anna")

print(result)



### Questions

1. What appears first on the screen?

## Hello Anna appears first.

---

2. What is stored in the variable `result`?

## None. Since the greet subprogram does not have a return statement, it  returns Python's special empty value.

---

3. Does the subprogram return a value?

## No, it does not explicitly return a value (acts as a procedure).

---



# Part 3. Identifying Functions and Procedures

Analyze each subprogram.



```python
def get_name():
    return input("Name: ")

def print_report():
    print("Report generated")

def calculate_average(a, b, c):
    return (a + b + c) / 3

def display_message():
    print("Welcome")
```

Complete the table.

| Subprogram        | Function or Procedure |
| ----------------- | --------------------- |
| get_name          |         Function              |
| print_report      |                Procedure       |
| calculate_average |               Function        |
| display_message   |          Procedure             |




# Part 4. Creating Procedures

## Guided Exercise

Complete the following procedure.




In [ ]:
def display_header():
    print("University")
    print("Structured Programming")
    print("Subprogram Practice")

display_header()


Expected output:

```text
University
Structured Programming
Subprogram Practice
```

---

## Challenge

Create a procedure called `display_table_of_5()` that prints the multiplication table of 5.

Expected output:

```text
5 x 1 = 5
5 x 2 = 10
...
5 x 10 = 50
```


# My procedure
```
# def display_table_of_5():
    for i in range(1, 11):
        print(f"5 x {i} = {5 * i}")

display_table_of_5()

display_header()
```



display_header()


# Part 5. Creating Functions

## Guided Exercise

Complete the following function.



In [ ]:
def square(number):
    return number * number

print(square(4))


16




Expected output:

```text
16
```



## Challenge

Create a function called `is_even(number)`.

The function must return:

* True if the number is even.
* False if the number is odd.

Test your function:

```python
print(is_even(8))
print(is_even(7))
```



In [ ]:
def is_even(number):
    if number % 2 == 0:
        return True
    else:
        return False

print(is_even(8))  # Returns True
print(is_even(7))  # Returns False



# Part 6. Introduction to Recursion

A recursive function is a function that calls itself.

Observe the following example.



In [ ]:
def countdown(n):

    print(n)

    if n > 1:
        countdown(n - 1)

countdown(5)



Execute the program.

### Questions

1. Which numbers are displayed?

## The numbers 5, 4, 3, 2, 1 are displayed

2. What happens when `n` becomes 1?

## The condition if n > 1: becomes false. Therefore, the function skips the recursive call and finishes executing without calling itself again.
---

3. Why does the program stop?

## It stops because it reaches the condition where n is no longer greater than 1, preventing any further recursive calls.
---



# Part 7. Understanding the Base Case

Every recursive function needs a condition that stops the recursion.

Observe the following example.



In [ ]:
def say_hello(n):

    if n == 0:
        return

    print("Hello")
    say_hello(n - 1)

say_hello(4)

Hello
Hello
Hello
Hello


### Questions

1. How many times is "Hello" printed?

## It is printed 4 times.
---

2. What is the base case?

## The base case is if n == 0:, which terminates the recursion by executing an empty return.

---

3. What would happen if the `if` statement were removed?

## The function would keep calling itself infinitely (say_hello(-1), say_hello(-2), etc.) until Python runs out of memory and crashes with an error: maximum recursion depth exceeded.
---


# Part 8. Building a Recursive Function

Complete the following function.



In [ ]:
def count_to_one(n):

    print(n)

    if n > 1:
        count_to_one(n - 1)

count_to_one(10)

Observe carefully how the function calls itself.

### Reflection

What is the role of the statement:

```python
count_to_one(n - 1)
```

*Answer*

This is the recursive call. It instructs the function to call a new instance of itself with a decremented value of **n** (one less than the current value), driving the sequence down toward the base case.
---


# Part 9. First Recursive Mathematical Example

The factorial of a number is defined as:

```text
4! = 4 × 3 × 2 × 1
```

Observe the following code.

```python
def factorial(n):

    if n == 1:
        return 1

    return n * factorial(n - 1)

print(factorial(4))
```

### Activity

Complete the table.

| Function Call | Waiting For |
| ------------- | ----------- |
| factorial(4)  |       4 * factorial(3)      |
| factorial(3)  |        3 * factorial(2)     |
| factorial(2)  |      2 * factorial(1)       |
| factorial(1)  |      Returns 1 directly       |

---

### Reflection

Why does the function eventually stop?

# The function stops because it eventually hits the base case if n == 1:, which immediately returns 1 without making another recursive call, allowing all the pending multiplications to resolve.

---

---

---

# Part 10. Final Challenge

Create a program that includes:

## Procedures

* display_menu()
* say_goodbye()

## Functions

* calculate_area()
* calculate_average()

## Recursive Function

* countdown()

The menu should allow the user to select an option.

```text
1. Calculate Area
2. Calculate Average
3. Countdown
4. Exit
```

### Requirements

* Use at least two procedures.
* Use at least two functions that return values.
* Use one recursive function.
* Organize the program into subprograms.


1. What is the difference between a function and a procedure?

The function executes a block of code and explicitly passes data back to the caller using a return statement. While the procedure executes a series of actions but does not pass a value back.
---

---

2. What is recursion?

# Recursion is a programming technique where a subprogram solves a problem by calling a smaller, simpler instance of itself.

---

---

3. Why does every recursive function need a base case?

A base case provides a termination condition that stops the function from calling itself. Without it, the function would loop infinitely and crash the program with a stack overflow error (RecursionError).
---

---

4. Which topic was most difficult for you?

# The most difficult topic was understanding recursion. It was tricky to visualize how a function can call itself multiple times and keep track of different values of n simultaneously without getting confused.

---

---

In [ ]:
#Part 10. Final Challenge (CONTAINS ADDITIONAL PRACTICE REQUIREMENTS)

# PROCEDURES

def display_menu():
    print("\n" + "="*25)
    print("1. Calculate Rectangle/Parallelogram Area")
    print("2. Calculate Average of Three Numbers")
    print("3. Recursive Countdown")
    print("4. Check if Number is Even or Odd")  # Menu Option
    print("5. Exit")
    print("="*25)

def say_goodbye():
    print("\nThank you for using the program. Goodbye!")


# FUNCTIONS WITH ROBUST VALIDATION

def get_valid_float(prompt, allow_negative=False):
    """Handles both error handling (type crashing) and input validation (logical values)."""
    while True:
        try:
            value = float(input(prompt))
            if not allow_negative and value <= 0:
                print("Error: The value must be greater than zero. Please try again.")
                continue
            return value
        except ValueError:
            print("Invalid input! Please enter a valid numerical value.")

def get_valid_int(prompt, minimum_val=None):
    """Validates integer inputs for options like countdown."""
    while True:
        try:
            value = int(input(prompt))
            if minimum_val is not None and value < minimum_val:
                print(f"Error: The value must be at least {minimum_val}.")
                continue
            return value
        except ValueError:
            print("Invalid input! Please enter a whole number.")

def calculate_area(base, height):
    return base * height

def calculate_average(num1, num2, num3):
    return (num1 + num2 + num3) / 3


# RECURSIVE FUNCTION

def countdown(n):
    print(n)
    if n > 1:
        countdown(n - 1)


# PROGRAM LOOP

def main():
    while True:
        display_menu()
        choice = input("Select an option (1-5): ").strip()

        if choice == '1':
            print("\n--- Calculate Area ---")
            b = get_valid_float("Enter the base length: ")
            h = get_valid_float("Enter the height: ")
            area = calculate_area(b, h)
            print(f"Success: A shape with a base of {b} and height of {h} has a total area of {area} square units.")

        elif choice == '2':
            print("\n--- Calculate Average ---")
            # Allows negative numbers since an average can be negative
            n1 = get_valid_float("Enter first number: ", allow_negative=True)
            n2 = get_valid_float("Enter second number: ", allow_negative=True)
            n3 = get_valid_float("Enter third number: ", allow_negative=True)
            avg = calculate_average(n1, n2, n3)
            print(f"Result: The mathematical average of [{n1}, {n2}, {n3}] is {avg:.2f}.")

        elif choice == '3':
            print("\n--- Countdown ---")
            start = get_valid_int("Enter a positive starting integer: ", minimum_val=1)
            print(f"Starting recursive countdown from {start}:")
            countdown(start)

        elif choice == '4':  # Additional menu option
            print("\n--- Even or Odd Checker ---")
            num = get_valid_int("Enter any integer to check: ")
            if num % 2 == 0:
                print(f"The number {num} is Even.")
            else:
                print(f"The number {num} is Odd.")

        elif choice == '5':
            say_goodbye()
            break

        else:
            # Error handling for invalid choice
            print("⚠️ Selection Error: 'Let's try that again.' Please choose a valid option from 1 to 5.")

main()


1. Calculate Rectangle/Parallelogram Area
2. Calculate Average of Three Numbers
3. Recursive Countdown
4. Check if Number is Even or Odd
5. Exit


# References

Downey, A. (2024). *Think Python: How to Think Like a Computer Scientist*. Green Tea Press.

Python Software Foundation. *Python Documentation*. Available at: https://docs.python.org/

Gaddis, T. (2023). *Starting Out with Python* (6th Edition). Pearson.


# Submission Instructions

To receive credit for this activity, complete the following steps:

1. Make a copy of the notebook and save it in your personal repository.

2. Complete all sections of the worksheet:

   * Run every example provided.
   * Answer all reflection questions.
   * Complete all guided exercises.
   * Complete all challenge activities.

3. Verify that all code cells execute correctly before submitting.

4. Once the notebook is finished, mark the assignment as **Completed** in Google Classroom.



# Additional Practice

You are encouraged to improve your solutions by adding:

* Input validation
* Error handling
* Additional menu options
* More descriptive output messages

These improvements are optional but highly recommended as a way to strengthen your programming skills and gain additional practice with structured programming concepts.
